Global optimization

petermattia opened the pull request:
While `scipy.optimize.curve_fit` is a great convenience function for fitting, the results are often sensitive to the initial conditions. EIS fitting is often prone to this issue given the fairly large number of parameters that are fit even for simple circuits. Global optimization algorithms attempt to avoid this issue. This PR adds the `scipy.optimize.basinhopping` global optimization algorithm as an option for fitting.

For example, using the sample data on [the Getting Started page](https://impedancepy.readthedocs.io/en/latest/getting-started.html#step-5-analyze-visualize-the-results) (also implemented in `test_circuit_fit`), `curve_fit` has an RMSE of 0.000413, while `basinhopping` has an RMSE of 0.000357. A comparison between the fitted parameters is below:

| Parameter | curve_fit | basinhopping |
|-----------|-----------|---------------|
| R_0       | 1.65e-2   | 1.65e-2       |
| R_1       | 8.68e-3   | 8.77e-3       |
| C_1       | 3.32      | 3.41          |
| R_2       | 5.39e-3   | 5.45e-3       |
| W_{o1,0}  | 6.31e-2   | 1.32e-1       |
| W_{o1,1}  | 2.33e2    | 1.10e3        |
| C_2       | 2.2e-1    | 2.23e-1       |

Note that `scipy.optimize` has a number of global optimization algorithms (see [here](https://docs.scipy.org/doc/scipy/reference/optimize.html#global-optimization)). However, `shgo` took many hours to converge for me, and the other three algorithms do not accept functions with infinite bounds. We can also consider implementing global optimization algorithms from other packages (e.g. [`mystic`](https://mystic.readthedocs.io/en/latest/)), but this at least is a start.

Some other changes:
- Added `test_circuit_fit` tests to `test_fitting.py`; tests both local and global optimization
- Added docstrings to some functions in `fitting.py`
- Renamed some single-letter variables (e.g. `f` -> `frequencies`, `ub` -> `upper_bounds`, etc)

petermattia commented:
Not sure why the build failed...`The command "flake8 impedance --exclude=*.ipynb_checkpoints" failed and exited with 1 during .`

BGerwe commented:
The TravisCI log says
```
flake8 impedance --exclude=*.ipynb_checkpoints
impedance/models/circuits/fitting.py:30:61: W291 trailing whitespace
impedance/models/circuits/fitting.py:35:37: W291 trailing whitespace
The command "flake8 impedance --exclude=*.ipynb_checkpoints" failed and exited with 1 during .```

So you have some trailing whitespaces on lines 30 and 35. You can also run this locally in a terminal in your impedance.py working directory.

petermattia commented:
Whoops, I made some small changes after running flake8 locally. Let's see if this works

petermattia commented:
I might have submitted this a tad early. I made two changes:
- Added a default seed
- Added try/except for computing `perror` (this caused the previous build to fail)

Btw, a really useful future PR would be to add a lot more tests to `test_fitting.py`

petermattia commented:
Hi @BGerwe and @mdmurbach, how does this look? Please lmk if there's anything else needed before merging. Thanks!

mdmurbach commented:
Awesome, I think this is a great feature to include and the fact that it doesn't change the api or add any dependencies is great.

I can take a look at the actual code later tonight, but at first glance everything looks good. Thanks for fixing those flake8 issues and adding a test.

mdmurbach commented on impedance/models/circuits/fitting.py:
```diff
@@ -66,37 +86,67 @@ def circuit_fit(frequencies, impedances, circuit, initial_guess, constants,
     # set upper and lower bounds on a per-element basis
     if bounds is None:
-        lb, ub = [], []
+        lower_bounds, upper_bounds = [], []
```
nice, I like the transition to better variable names here

mdmurbach commented on impedance/models/circuits/fitting.py:
```diff
@@ -66,37 +86,67 @@ def circuit_fit(frequencies, impedances, circuit, initial_guess, constants,
+        try:
+            perror = inv(np.dot(jac.T, jac)) * opt_function(popt) ** 2
+        except np.linalg.LinAlgError:
+            print('Failed to compute perror')
```
I've found that it's much easier to deal with messages like this if they are warnings (you can choose to ignore them, raise exceptions on them, etc. which can be useful, especially in automated analysis) rather than prints, so can we change this to `warnings.warn("Failed to compute perror")`? You'll also have to `import warnings` at the top of the file

BGerwe replied:
Ooh, that's a great tip! This would be a separate ticket, but I think we should change the printing behavior in `BaseCircuit.predict()` to this. 
I was recently predicting impedance within a regression and it printed "Simulating circuit based on initial parameters" like 500 times 😆

petermattia replied:
Good call! Changed, also slightly rearranged imports per [PEP8](https://www.python.org/dev/peps/pep-0008/#imports)

mdmurbach replied:
Haha, yeah @BGerwe that's exactly the printing issue that drove my original comment...

mdmurbach replied:
PR to update the other print to warnings here: https://github.com/ECSHackWeek/impedance.py/pull/158

mdmurbach commented on impedance/models/circuits/fitting.py:
```diff
@@ -247,6 +297,19 @@ def count_parens(string):
 
 
 def extract_circuit_elements(circuit):
+    """ Extracts circuit elements from a circuit string.
```
🚀 thanks for documenting

mdmurbach commented:
Overall, looks great. Only one small suggestion on using warnings instead of printing, but otherwise looks good. I'll plan on releasing a new version once we get this merged in as well 🚀 thanks @petermattia!

mdmurbach reviewed (changes requested):
Overall, looks great. Only one small suggestion on using warnings instead of printing, but otherwise looks good. I'll plan on releasing a new version once we get this merged in as well 🚀 thanks @petermattia!

petermattia commented:
Ok, done! Also caught a bug in testing, now the seed actually does something 😅

mdmurbach commented:
@all-contributors please add @petermattia for tests, docs
